---
title: "Introduction to Genomica"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to Genomica}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(Genomica)
```

## Welcome to Genomica

Genomica provides a suite for the analysis of complex genomic data using a false discovery rate corrected, linear mixed model approach, whilst also allowing to test for interactions between two predictors.

These data can be a collection of genes or microbial taxa (i.e., generically called features hereafter), which could be both from prokaryotic and eucaryotic experiments, such as for example taxonomic or functional-gene/AMR tables derived from shotgun metagenomics or metatranscriptomics, or quantitative gene tables from qPCR experiments. The only inputs required by the user are two data frames (Data and Metadata) and, if required, Genomica will carry out the log10 transformation of pre-normalised data.
The outputs from Genomica are summarised in two .xlsx files describing the whole list of statistical results per feature and a list of significant comparisons, respectively.
The tests carried out in Genomica develop through linear mixed models (LMM) via calling the package lmer4 (Bates et al., 2015) and calculating the P-value via type III ANOVA using the Satterthwaite's method through the package “lmerTest” (Kuznetsova et al., 2017).
Moreover, the tests, and ultimately the significance levels are based on the calculation of the false discovery rate (type I error) under repeated testing. This is carried out via calling the package fuzzySim (A. Marcia Barbosal, 2015), which uses the Benjamini & Hochberg correction to generate p.adjust values based on the p values generated by the LMM.

## Data frame preparation

In order to carry out the analyses, you will need to load two data frames (i.e., Data and Metadata). In this section, some suggestions are provided in order to prepare the data frames for the analyses.

In Genomica, you will find two pre-loaded demo data frames, i.e., Data_Demo (3 taxonomical TSS normalised features across 42 samples) and Metadata_Demo (informing on the study layout, such as treatment and random factor allocation relative to the 42 samples). Hereafter, we will use these demo data frames for tutorial purposes. To call these data frames simply load the package via using:

```{r}
library(Genomica)
```
and then you can store the demo data frames into variables, by typing, for example:
```{r}
Data<-Data_Demo
Metadata<-Metadata_Demo
```
Importantly, Data_Demo is a subset of a bigger data frame, therefore the sum of the features per sample is not equal to 100 (%).

### Data

IMPORTANT, the data frame "Data" must be formatted with features as rows and samples as columns, such as for example as we can see for the Data_Demo data frame:

```{r}
print(Data[,1:5])
```

This is the most common format output of, for example, metagenomic pipelines, however at times your Data could be generated in the opposite way (i.e., with features as columns), like in the below example:

```{r}
print(Wrong_Format_Data[1:5,])
```

In such cases, you must transpose your Data it into the appropriate format prior to carrying out the analysis with Genomica. This can be done as follows:

```{r}
Corrected_Format_Data<-as.data.frame(t(Wrong_Format_Data))
```
which would generate:
``` {r}
print(Corrected_Format_Data[,1:5])
```

IMPORTANT: The features names in Data must be assigned as rows (row names), whilst the sample name must be assigned as columns (column names).
If your data frame is embedded with features as a separate column, such as:
```{r}
print(No_Features_As_Rows_Data[,1:5])
```
you must assign the rownames and delete the feature-embedded column. This can be done as follows:
```{r}
rownames(No_Features_As_Rows_Data)<-No_Features_As_Rows_Data$Feature
No_Features_As_Rows_Data<-No_Features_As_Rows_Data[,-1]
```
The data frame could then be renamed:
```{r}
Feature_As_Rows_Data<-No_Features_As_Rows_Data
```
Whose first five columns would look like the below:
```{r}
print(Feature_As_Rows_Data[,1:5])
```
At this point, your Data is ready for downstream analyses with Genomica.

### Metadata
Your metadata must be formatted with samples as rows and features as columns (opposite to what seen for Data).
IMPORTANT: the sample names must be assigned as rows (row names), please refer to the above instructions on how to achieve this.
An example of a metadata file can be found within the demo metadata file found in Genomica (we are only looking at the first five rows in the below example):
```{r}
print(Metadata_Demo[1:5,])
```
## Performing an analysis with Genomica
Genomica carries out a linear mix model on all the features in your Data, thus it will further analyse the significant comparisons (according to the false discovery rate) and provide further information on the eventual significant comparisons within the specified predictor.
IMPORTANT, Genomica will not carry out the linear mixed model of the feature whose quantity is 0 throughout all the samples in the data frame.

Currently, Genomica allows to perform the analysis either with a single or with two predictors, and in the latter case, it allows to test for the interaction between the two predictors.

### One predictor

In case your study layout present one predictor only, you should use the function "genomica_one_predictor()" for the analysis.
The below is an example on how to set up the function, when using the demo data frames:

Load the data frames (Data_Demo and Metadata_Demo, in this case)
```{r}
Data<-Data_Demo
Metadata<-Metadata_Demo
```
As seen above, in this example Data is a data frame of 3 taxonomical features (rows) across 42 samples (columns), and below the first five columns of Data are shown:
```{r}
print(Data[,1:5])
```
In parallel, Metadata is a data frame with 42 rows (samples) and 5 columns (Sample ID, Treatment, Sex, Room and Block).
The column Treatment contains the treatment layout (Control, T1 or T2) for this study. This will be the main predictor for this analysis, and the levels for this predictor (P_Levels) are c(Control, T1 or T2).
The column Room contains the random effect of the model, with levels (R1_Levels) set to c(Room1 and Room2).
The first five rows of Metadata are shown below:
```{r}
print(Metadata[1:5,])
```
The argument "Already_Log10_transformed" in "genomica_one_predictor()" will inform on whether the Log10 transformation had been previously carried out on the data set. In our case, the features in Data are TSS  normalised but non transformed, hence we will include "Already_Log10_transformed = c('NO')" within the function.

Finally, through the argument "Folder_Name", you need to provide a string to be attached to the output folder, in the example below this is specified via "Folder_Name=c('Test)", which will result in the folder name being "Genomica_Output_Test".

IMPORTANT. When you run a new analysis, Genomica will automatically delete all the previous files present in the output folder (if already present), so it is pivotal that you set a new folder name if you don't want to permanently lose the previous results.
```{r}
genomica_one_predictor(Data = Data,Metadata = Metadata,
Predictor = c('Treatment'),P_Levels=c('Control','T1','T2'),
R_Effects=c('Room'),R1_Levels=c('Room1','Room2','Room3','Room4','Room5','Room6','Room7'),
Already_Log10_transformed = c('NO'),Folder_Name=c('Test'))
```
In this particular case (with only 3 features) no significant (p.adjust) associations are found by Genomica, thus the only output is a .xlsx file containing the results of all the linear models carried out (3 in this case). This file can be found in the folder "Genomica_Output_Test" automatically created by the function.

IMPORTANT: the predictor could be a numerical variable (e.g., methane emission) instead of a categorical one, in which case the concept of levels would be meaningless. To set up the levels for numerical variables please state:
```{r}
#P_Levels=c(0)
```
within genomica_one_predictor().

### Two predictors

To test the main effect of two predictors (e.g., Treatment and Sex) and their interaction, you should use the function "genomica_two_predictors()"

Following the structure of the previous example, let's load the two data frames as we did before:

```{r}
Data<-Data_Demo
Metadata<-Metadata_Demo
```

As seen above, in Metadata there are two columns for the two different predictors (i.e., Treatment and Sex).
As per above, "Treatment" will be the main predictor (Predictor1) for this example, with levels (P1_Levels) being c(Control, T1 or T2).

The second predictor  (Predictor2) for this example will be "Sex", whose levels (Female and Male) will be set via the argument "P2_Levels" in "genomica_two_predictors()".

The column Room contains the random effect of the model, with levels (R1_Levels) set to c(Room1 and Room2).

The first five rows of Metadata are shown below:

```{r}
print(Metadata[1:5,])
```
As per above, we should include:
```{r}
#Already_Log10_transformed = c("NO")
```
within the function.

Finally, as seen above you need to provide a string to be attached to the output folder, in the example below this is specified via "Folder_Name=c('Test)", which will result in the folder name being "Genomica_Output_Two_Predictors_Test".

IMPORTANT. When you run a new analysis, Genomica will automatically delete all the previous files present in the output folder (if already present), so it is pivotal that you set a new folder name if you don't want to permanently lose the previous results.

At this point we can call the function as follows:
```{r}
genomica_two_predictors(Data = Data,Metadata = Metadata,
                        Predictor1= c('Treatment'),
                        Predictor2=c('Sex'),
                        P1_Levels=c('Control','T1','T2'),
                        P2_Levels=c('Female','Male'),
                        R_Effects=c('Room'),
                        R1_Levels=c('Room1','Room2','Room3','Room4','Room5','Room6','Room7'),
                        Already_Log10_transformed = c('No'),Folder_Name=c('Test'))
```
And as per above no significant associations are found in this particular example, however a full summary is retained in the .xlsx file in the folder "Genomica_Output_Two_Predictors_Test".

IMPORTANT: one or both predictors could be a numerical variable (e.g., methane emission and body weight), in which case please state:
```{r}
#P1_Levels = c(0),
#P2_Levels = c(0),
```
within genomica_two_predictors().

## References

Bates, D., Mächler, M., Bolker, B., & Walker, S. (2015). Fitting Linear Mixed-Effects Models Using lme4. Journal of Statistical Software, 67(1), 1–48. https://doi.org/10.18637/jss.v067.i01

Kuznetsova, A., Brockhoff, P. B., & Christensen, R. H. B. (2017). lmerTest Package: Tests in Linear Mixed Effects Models. Journal of Statistical Software, 82(13), 1–26. https://doi.org/10.18637/jss.v082.i13

Barbosa, A.M. (2015), fuzzySim: applying fuzzy logic to binary similarity indices in ecology. Methods Ecol Evol, 6: 853-858. https://doi.org/10.1111/2041-210X.12372
